Business Intelligence Project    
Name: Zehra Ahmed, Farah Inayat, Kisa Fatima, Zuha Aqib    
Date: 18-May-2025

In [736]:
# print when the last code was run
from datetime import datetime
datetime.now().strftime("%Y-%m-%d %H:%M:%S")

'2025-05-28 15:14:07'

# Imports

In [737]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from datetime import datetime, timedelta

### Some Functions
some preliminary functions we can use in the code

In [738]:
# Function to get current date and time as a string
def get_current_datetime():
    return datetime.now().strftime("%Y-%m-%d %H:%M:%S")

# Data Loading

In [739]:
# Load the dataset
df = pd.read_csv('../data/flights_sample_3m.csv')

pd.set_option('display.max_columns', None)

In [740]:
df

FL_DATE                 AIRLINE                 AIRLINE_DOT  \
0        2019-01-09   United Air Lines Inc.   United Air Lines Inc.: UA   
1        2022-11-19    Delta Air Lines Inc.    Delta Air Lines Inc.: DL   
2        2022-07-22   United Air Lines Inc.   United Air Lines Inc.: UA   
3        2023-03-06    Delta Air Lines Inc.    Delta Air Lines Inc.: DL   
4        2020-02-23        Spirit Air Lines        Spirit Air Lines: NK   
...             ...                     ...                         ...   
2999995  2022-11-13  American Airlines Inc.  American Airlines Inc.: AA   
2999996  2022-11-02  American Airlines Inc.  American Airlines Inc.: AA   
2999997  2022-09-11    Delta Air Lines Inc.    Delta Air Lines Inc.: DL   
2999998  2019-11-13        Republic Airline        Republic Airline: YX   
2999999  2019-06-15  Southwest Airlines Co.  Southwest Airlines Co.: WN   

        AIRLINE_CODE  DOT_CODE  FL_NUMBER ORIGIN          ORIGIN_CITY DEST  \
0                 UA     19977       1562    FLL  Fort Lauderdale, FL  EWR   
1                 DL     19790       1149    MSP      Minneapolis, MN  SEA   
2                 UA     19977        459    DEN           Denver, CO  MSP   
3                 DL     19790       2295    MSP      Minneapolis, MN  SFO   
4                 NK     20416        407    MCO          Orlando, FL  DFW   
...              ...       ...        ...    ...                  ...  ...   
2999995           AA     19805       1522    JAX     Jacksonville, FL  CLT   
2999996           AA     19805       1535    ORD          Chicago, IL  AUS   
2999997           DL     19790       2745    HSV       Huntsville, AL  ATL   
2999998           YX     20452       6134    BOS           Boston, MA  LGA   
2999999           WN     19393       2823    LGB       Long Beach, CA  SJC   

                     DEST_CITY  CRS_DEP_TIME  DEP_TIME  DEP_DELAY  TAXI_OUT  \
0                   Newark, NJ          1155    1151.0       -4.0      19.0   
1                  Seattle, WA          2120    2114.0       -6.0       9.0   
2              Minneapolis, MN           954    1000.0        6.0      20.0   
3            San Francisco, CA          1609    1608.0       -1.0      27.0   
4        Dallas/Fort Worth, TX          1840    1838.0       -2.0      15.0   
...                        ...           ...       ...        ...       ...   
2999995          Charlotte, NC          1742    1740.0       -2.0      10.0   
2999996             Austin, TX          1300    1254.0       -6.0      10.0   
2999997            Atlanta, GA           534     615.0       41.0      16.0   
2999998           New York, NY          1600    1555.0       -5.0      19.0   
2999999           San Jose, CA           730     727.0       -3.0       9.0   

         WHEELS_OFF  WHEELS_ON  TAXI_IN  CRS_ARR_TIME  ARR_TIME  ARR_DELAY  \
0            1210.0     1443.0      4.0          1501    1447.0      -14.0   
1            2123.0     2232.0     38.0          2315    2310.0       -5.0   
2            1020.0     1247.0      5.0          1252    1252.0        0.0   
3            1635.0     1844.0      9.0          1829    1853.0       24.0   
4            1853.0     2026.0     14.0          2041    2040.0       -1.0   
...             ...        ...      ...           ...       ...        ...   
2999995      1750.0     1845.0      6.0          1907    1851.0      -16.0   
2999996      1304.0     1514.0      5.0          1556    1519.0      -37.0   
2999997       631.0      759.0      6.0           729     805.0       36.0   
2999998      1614.0     1704.0      8.0          1728    1712.0      -16.0   
2999999       736.0      828.0      2.0           840     830.0      -10.0   

         CANCELLED CANCELLATION_CODE  DIVERTED  CRS_ELAPSED_TIME  \
0              0.0               NaN       0.0             186.0   
1              0.0               NaN       0.0             235.0   
2              0.0               NaN       0.0             118.0   
3              0

In [741]:
df.shape

(3000000, 32)

# Data Preparation

## Data Statistics

In [742]:
# Check for missing values
print("Missing values per column:")
df.isnull().sum()

Missing values per column:


FL_DATE                          0
AIRLINE                          0
AIRLINE_DOT                      0
AIRLINE_CODE                     0
DOT_CODE                         0
FL_NUMBER                        0
ORIGIN                           0
ORIGIN_CITY                      0
DEST                             0
DEST_CITY                        0
CRS_DEP_TIME                     0
DEP_TIME                     77615
DEP_DELAY                    77644
TAXI_OUT                     78806
WHEELS_OFF                   78806
WHEELS_ON                    79944
TAXI_IN                      79944
CRS_ARR_TIME                     0
ARR_TIME                     79942
ARR_DELAY                    86198
CANCELLED                        0
CANCELLATION_CODE          2920860
DIVERTED                         0
CRS_ELAPSED_TIME                14
ELAPSED_TIME                 86198
AIR_TIME                     86198
DISTANCE                         0
DELAY_DUE_CARRIER          2466137
DELAY_DUE_WEATHER   

In [743]:
# Display data types
print("\nData types:")
df.dtypes


Data types:


FL_DATE                     object
AIRLINE                     object
AIRLINE_DOT                 object
AIRLINE_CODE                object
DOT_CODE                     int64
FL_NUMBER                    int64
ORIGIN                      object
ORIGIN_CITY                 object
DEST                        object
DEST_CITY                   object
CRS_DEP_TIME                 int64
DEP_TIME                   float64
DEP_DELAY                  float64
TAXI_OUT                   float64
WHEELS_OFF                 float64
WHEELS_ON                  float64
TAXI_IN                    float64
CRS_ARR_TIME                 int64
ARR_TIME                   float64
ARR_DELAY                  float64
CANCELLED                  float64
CANCELLATION_CODE           object
DIVERTED                   float64
CRS_ELAPSED_TIME           float64
ELAPSED_TIME               float64
AIR_TIME                   float64
DISTANCE                   float64
DELAY_DUE_CARRIER          float64
DELAY_DUE_WEATHER   

In [744]:
# Get summary statistics
print("\nSummary statistics:")
df.describe(include='all')


Summary statistics:


FL_DATE                 AIRLINE                 AIRLINE_DOT  \
count      3000000                 3000000                     3000000   
unique        1704                      18                          18   
top     2019-07-25  Southwest Airlines Co.  Southwest Airlines Co.: WN   
freq          2379                  576470                      576470   
mean           NaN                     NaN                         NaN   
std            NaN                     NaN                         NaN   
min            NaN                     NaN                         NaN   
25%            NaN                     NaN                         NaN   
50%            NaN                     NaN                         NaN   
75%            NaN                     NaN                         NaN   
max            NaN                     NaN                         NaN   

       AIRLINE_CODE      DOT_CODE     FL_NUMBER   ORIGIN  ORIGIN_CITY  \
count       3000000  3.000000e+06  3.000000e+06  3000000      3000000   
unique           18           NaN           NaN      380          373   
top              WN           NaN           NaN      ATL  Chicago, IL   
freq         576470           NaN           NaN   153556       157368   
mean            NaN  1.997629e+04  2.511536e+03      NaN          NaN   
std             NaN  3.772846e+02  1.747258e+03      NaN          NaN   
min             NaN  1.939300e+04  1.000000e+00      NaN          NaN   
25%             NaN  1.979000e+04  1.051000e+03      NaN          NaN   
50%             NaN  1.993000e+04  2.152000e+03      NaN          NaN   
75%             NaN  2.036800e+04  3.797000e+03      NaN          NaN   
max             NaN  2.045200e+04  9.562000e+03      NaN          NaN   

           DEST    DEST_CITY  CRS_DEP_TIME      DEP_TIME     DEP_DELAY  \
count   3000000      3000000  3.000000e+06  2.922385e+06  2.922356e+06   
unique      380          373           NaN           NaN           NaN   
top         ATL  Chicago, IL           NaN           NaN           NaN   
freq     153569       158087           NaN           NaN           NaN   
mean        NaN          NaN  1.327062e+03  1.329776e+03  1.012333e+01   
std         NaN          NaN  4.858789e+02  4.993101e+02  4.925183e+01   
min         NaN          NaN  1.000000e+00  1.000000e+00 -9.000000e+01   
25%         NaN          NaN  9.150000e+02  9.160000e+02 -6.000000e+00   
50%         NaN          NaN  1.320000e+03  1.323000e+03 -2.000000e+00   
75%         NaN          NaN  1.730000e+03  1.739000e+03  6.000000e+00   
max         NaN          NaN  2.359000e+03  2.400000e+03  2.966000e+03   

            TAXI_OUT    WHEELS_OFF     WHEELS_ON       TAXI_IN  CRS_ARR_TIME  \
count   2.921194e+06  2.921194e+06  2.920056e+06  2.920056e+06  3.000000e+06   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean    1.664305e+01  1.352361e+03  1.462500e+03  7.678982e+00  1.490561e+03   
std     9.192901e+00  5.008727e+02  5.272368e+02  6.269639e+00  5.115476e+02   
min     1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%     1.100000e+01  9.310000e+02  1.049000e+03  4.000000e+00  1.107000e+03   
50%     1.400000e+01  1.336000e+03  1.501000e+03  6.000000e+00  1.516000e+03   
75%     1.900000e+01  1.752000e+03  1.908000e+03  9.000000e+00  1.919000e+03   
max     1.840000e+02  2.400000e+03  2.400000e+03  2.490000e+02  2.400000e+03   

            ARR_TIME     ARR_DELAY     CANCELLED CANCELLATION_CODE  \
count   2.920058e+06  2.913802e+06  3.000000e+06             79140   
unique           NaN           NaN           NaN                 4   
top              NaN           NaN           NaN                 B   
freq             NaN           NaN           NaN             28772   
mean    1.466511e+03  4.260858e+00  2.638000e

## Changing Data Type

### Defined Functions
here are some functions we can use for changing the datatype

In [745]:
def to_string(df, column_name):
    """
    Convert a pandas DataFrame column to a string.
    """
    df[column_name] = df[column_name].astype('string')
    print(f"Column Name: '{column_name}', Type: {df[column_name].dtype}")
    return df  

In [746]:
def to_date(df, column_name, format=None):
    """
    Convert a pandas DataFrame column to datetime.
    """
    df[column_name] = pd.to_datetime(df[column_name], format=format)
    print(f"Column Name: '{column_name}', Type: {df[column_name].dtype}")
    return df

In [747]:
def to_int(df, column_name):
    """
    Convert a pandas DataFrame column to a int.
    """
    df[column_name] = df[column_name].astype('int64')
    print(f"Column Name: '{column_name}', Type: {df[column_name].dtype}")
    return df  

In [748]:
def to_float(df, column_name):
    """
    Convert a pandas DataFrame column to a float.
    """
    df[column_name] = df[column_name].astype('float64')
    print(f"Column Name: '{column_name}', Type: {df[column_name].dtype}")
    return df  

### Implementation
so there is a date column lets make it a type date column, and there is a few columns which are object datatype, lets change to string for better readability

In [749]:
df = to_date(df, 'FL_DATE', '%Y-%m-%d')

Column Name: 'FL_DATE', Type: datetime64[ns]


In [750]:
df = to_string(df, 'AIRLINE')
df = to_string(df, 'AIRLINE_DOT')
df = to_string(df, 'AIRLINE_CODE')
df = to_string(df, 'ORIGIN')
df = to_string(df, 'ORIGIN_CITY')
df = to_string(df, 'DEST')
df = to_string(df, 'DEST_CITY')
df = to_string(df, 'CANCELLATION_CODE')

Column Name: 'AIRLINE', Type: string
Column Name: 'AIRLINE_DOT', Type: string
Column Name: 'AIRLINE_CODE', Type: string
Column Name: 'ORIGIN', Type: string
Column Name: 'ORIGIN_CITY', Type: string
Column Name: 'DEST', Type: string
Column Name: 'DEST_CITY', Type: string
Column Name: 'CANCELLATION_CODE', Type: string


## Fixing columns saath saath
so for each column we're just gonna check is it coherenet? are there any errors like mistypes? this is for only those columns which are text and have some textual information and their unique values are small and checkable

### AIRLINE
here lets extract its 18 unique values and check it

In [751]:
# Just checking unique airline names
print(df['AIRLINE'].value_counts())

AIRLINE
Southwest Airlines Co.                576470
Delta Air Lines Inc.                  395239
American Airlines Inc.                383106
SkyWest Airlines Inc.                 343737
United Air Lines Inc.                 254504
Republic Airline                      143107
Envoy Air                             121256
JetBlue Airways                       112844
Endeavor Air Inc.                     112463
PSA Airlines Inc.                     107050
Alaska Airlines Inc.                  100467
Spirit Air Lines                       95711
Mesa Airlines Inc.                     65012
Frontier Airlines Inc.                 64466
Allegiant Air                          52738
Hawaiian Airlines Inc.                 32114
Horizon Air                            20634
ExpressJet Airlines LLC d/b/a aha!     19082
Name: count, dtype: Int64


so there is no typo. lets move ahead to the next one

### AIRLINE DOT

In [752]:
print(df['AIRLINE_DOT'].value_counts())

AIRLINE_DOT
Southwest Airlines Co.: WN                576470
Delta Air Lines Inc.: DL                  395239
American Airlines Inc.: AA                383106
SkyWest Airlines Inc.: OO                 343737
United Air Lines Inc.: UA                 254504
Republic Airline: YX                      143107
Envoy Air: MQ                             121256
JetBlue Airways: B6                       112844
Endeavor Air Inc.: 9E                     112463
PSA Airlines Inc.: OH                     107050
Alaska Airlines Inc.: AS                  100467
Spirit Air Lines: NK                       95711
Mesa Airlines Inc.: YV                     65012
Frontier Airlines Inc.: F9                 64466
Allegiant Air: G4                          52738
Hawaiian Airlines Inc.: HA                 32114
Horizon Air: QX                            20634
ExpressJet Airlines LLC d/b/a aha!: EV     19082
Name: count, dtype: Int64


nothing again

### AIRLINE CODE

In [753]:
print(df['AIRLINE_CODE'].value_counts())

AIRLINE_CODE
WN    576470
DL    395239
AA    383106
OO    343737
UA    254504
YX    143107
MQ    121256
B6    112844
9E    112463
OH    107050
AS    100467
NK     95711
YV     65012
F9     64466
G4     52738
HA     32114
QX     20634
EV     19082
Name: count, dtype: Int64


nothing again but the column `AIRLINE_DOT` is very similar to the columns we already have - its just a concatentation of `AIRLINES` AND `AIRLINE_CODE`. So lets drop it.

In [754]:
df = df.drop('AIRLINE_DOT', axis=1)

In [755]:
df

FL_DATE                 AIRLINE AIRLINE_CODE  DOT_CODE  FL_NUMBER  \
0       2019-01-09   United Air Lines Inc.           UA     19977       1562   
1       2022-11-19    Delta Air Lines Inc.           DL     19790       1149   
2       2022-07-22   United Air Lines Inc.           UA     19977        459   
3       2023-03-06    Delta Air Lines Inc.           DL     19790       2295   
4       2020-02-23        Spirit Air Lines           NK     20416        407   
...            ...                     ...          ...       ...        ...   
2999995 2022-11-13  American Airlines Inc.           AA     19805       1522   
2999996 2022-11-02  American Airlines Inc.           AA     19805       1535   
2999997 2022-09-11    Delta Air Lines Inc.           DL     19790       2745   
2999998 2019-11-13        Republic Airline           YX     20452       6134   
2999999 2019-06-15  Southwest Airlines Co.           WN     19393       2823   

        ORIGIN          ORIGIN_CITY DEST              DEST_CITY  CRS_DEP_TIME  \
0          FLL  Fort Lauderdale, FL  EWR             Newark, NJ          1155   
1          MSP      Minneapolis, MN  SEA            Seattle, WA          2120   
2          DEN           Denver, CO  MSP        Minneapolis, MN           954   
3          MSP      Minneapolis, MN  SFO      San Francisco, CA          1609   
4          MCO          Orlando, FL  DFW  Dallas/Fort Worth, TX          1840   
...        ...                  ...  ...                    ...           ...   
2999995    JAX     Jacksonville, FL  CLT          Charlotte, NC          1742   
2999996    ORD          Chicago, IL  AUS             Austin, TX          1300   
2999997    HSV       Huntsville, AL  ATL            Atlanta, GA           534   
2999998    BOS           Boston, MA  LGA           New York, NY          1600   
2999999    LGB       Long Beach, CA  SJC           San Jose, CA           730   

         DEP_TIME  DEP_DELAY  TAXI_OUT  WHEELS_OFF  WHEELS_ON  TAXI_IN  \
0          1151.0       -4.0      19.0      1210.0     1443.0      4.0   
1          2114.0       -6.0       9.0      2123.0     2232.0     38.0   
2          1000.0        6.0      20.0      1020.0     1247.0      5.0   
3          1608.0       -1.0      27.0      1635.0     1844.0      9.0   
4          1838.0       -2.0      15.0      1853.0     2026.0     14.0   
...           ...        ...       ...         ...        ...      ...   
2999995    1740.0       -2.0      10.0      1750.0     1845.0      6.0   
2999996    1254.0       -6.0      10.0      1304.0     1514.0      5.0   
2999997     615.0       41.0      16.0       631.0      759.0      6.0   
2999998    1555.0       -5.0      19.0      1614.0     1704.0      8.0   
2999999     727.0       -3.0       9.0       736.0      828.0      2.0   

         CRS_ARR_TIME  ARR_TIME  ARR_DELAY  CANCELLED CANCELLATION_CODE  \
0                1501    1447.0      -14.0        0.0              <NA>   
1                2315    2310.0       -5.0        0.0              <NA>   
2                1252    1252.0        0.0        0.0              <NA>   
3                1829    1853.0       24.0        0.0              <NA>   
4                2041    2040.0       -1.0        0.0              <NA>   
...               ...       ...        ...        ...               ...   
2999995          1907    1851.0      -16.0        0.0              <NA>   
2999996          1556    1519.0      -37.0        0.0              <NA>   
2999997           729     805.0       36.0        0.0              <NA>   
2999998          1728    1712.0      -16.0        0.0              <NA>   
2999999           840     830.0      -10.0        0.0              <NA>   

         DIVERTED  CRS_ELAPSED_TIME  ELAPSED_TIME  AIR_TIME  DISTANCE  \
0             0.0             186.0         176.0     153.0    1065.0   
1             0.0             235.0         236.0     189.0    1399.0   
2             0.0             118.0         112.0      87.0     680.

### DOT_CODE
now lets explore dot_code to see what its values are like

In [756]:
df['DOT_CODE'].value_counts()

DOT_CODE
19393    576470
19790    395239
19805    383106
20304    343737
19977    254504
20452    143107
20398    121256
20409    112844
20363    112463
20397    107050
19930    100467
20416     95711
20378     65012
20436     64466
20368     52738
19690     32114
19687     20634
20366     19082
Name: count, dtype: int64

okay so its our intutiton that its related to the airline. according to the html dictionary - IT IS. so we can use that to our advantage - any cell that is empty, we can just find the related airline and extract its airline code and fill with that value. so we will do this in the MISSING VALUES section. lets see if we have any missing values:

In [757]:
df['DOT_CODE'].isnull().sum()

0

ok so we don't have missing values. thats clear.

# Missing Values
here we handle missing values for each column. 

In [758]:
df.shape

(3000000, 31)

In [759]:
df.isnull().sum()

FL_DATE                          0
AIRLINE                          0
AIRLINE_CODE                     0
DOT_CODE                         0
FL_NUMBER                        0
ORIGIN                           0
ORIGIN_CITY                      0
DEST                             0
DEST_CITY                        0
CRS_DEP_TIME                     0
DEP_TIME                     77615
DEP_DELAY                    77644
TAXI_OUT                     78806
WHEELS_OFF                   78806
WHEELS_ON                    79944
TAXI_IN                      79944
CRS_ARR_TIME                     0
ARR_TIME                     79942
ARR_DELAY                    86198
CANCELLED                        0
CANCELLATION_CODE          2920860
DIVERTED                         0
CRS_ELAPSED_TIME                14
ELAPSED_TIME                 86198
AIR_TIME                     86198
DISTANCE                         0
DELAY_DUE_CARRIER          2466137
DELAY_DUE_WEATHER          2466137
DELAY_DUE_NAS       

## DEP_TIME

In [760]:
df['DEP_TIME'].isnull().sum()

77615

In [761]:
df['DEP_TIME'].value_counts()

DEP_TIME
555.0    7991
557.0    7186
556.0    7184
655.0    6662
558.0    6640
         ... 
406.0       4
349.0       4
408.0       3
434.0       3
415.0       3
Name: count, Length: 1440, dtype: int64

so we see that all the missing values are either 77615 or more. we have an intuition that maybe those rows are those rows that have ALL the missing values. so we get all the missing values which are 77615 >= and extract their rows and see how many are extracted. if they are 77615, that means all those rows have majority columns empty - they have the important columns empty

In [762]:
null_rows = df[
    df['DEP_TIME'].isnull() & 
    df['DEP_DELAY'].isnull() & 
    df['TAXI_OUT'].isnull() & 
    df['WHEELS_OFF'].isnull() & 
    df['WHEELS_ON'].isnull() & 
    df['TAXI_IN'].isnull() & 
    df['ARR_TIME'].isnull() & 
    df['ARR_DELAY'].isnull() & 
    df['ELAPSED_TIME'].isnull() &
    df['AIR_TIME'].isnull() &
    df['DELAY_DUE_CARRIER'].isnull() &
    df['DELAY_DUE_WEATHER'].isnull() &
    df['DELAY_DUE_NAS'].isnull() &
    df['DELAY_DUE_SECURITY'].isnull() &
    df['DELAY_DUE_LATE_AIRCRAFT'].isnull()
]

null_rows

FL_DATE                 AIRLINE AIRLINE_CODE  DOT_CODE  FL_NUMBER  \
14      2020-04-07  Southwest Airlines Co.           WN     19393        687   
45      2022-01-02        Republic Airline           YX     20452       3467   
80      2020-04-20  Southwest Airlines Co.           WN     19393        246   
191     2021-12-31    Alaska Airlines Inc.           AS     19930       1097   
236     2020-04-14  Southwest Airlines Co.           WN     19393        446   
...            ...                     ...          ...       ...        ...   
2999779 2019-06-17       PSA Airlines Inc.           OH     20397       5305   
2999865 2021-10-22   SkyWest Airlines Inc.           OO     20304       3260   
2999910 2020-04-08           Allegiant Air           G4     20368       1595   
2999943 2020-03-30  American Airlines Inc.           AA     19805       1877   
2999946 2023-07-09        Republic Airline           YX     20452       4296   

        ORIGIN          ORIGIN_CITY DEST        DEST_CITY  CRS_DEP_TIME  \
14         SJC         San Jose, CA  LAX  Los Angeles, CA          2155   
45         ORD          Chicago, IL  DCA   Washington, DC          1935   
80         BNA        Nashville, TN  ATL      Atlanta, GA          1055   
191        IAD       Washington, DC  LAX  Los Angeles, CA          1720   
236        LAS        Las Vegas, NV  HOU      Houston, TX           500   
...        ...                  ...  ...              ...           ...   
2999779    DCA       Washington, DC  CAK        Akron, OH          1959   
2999865    ORD          Chicago, IL  MEM      Memphis, TN          1530   
2999910    PGD      Punta Gorda, FL  STC    St. Cloud, MN          1504   
2999943    FLL  Fort Lauderdale, FL  CLT    Charlotte, NC          1443   
2999946    BOS           Boston, MA  JFK     New York, NY          1734   

         DEP_TIME  DEP_DELAY  TAXI_OUT  WHEELS_OFF  WHEELS_ON  TAXI_IN  \
14            NaN        NaN       NaN         NaN        NaN      NaN   
45            NaN        NaN       NaN         NaN        NaN      NaN   
80            NaN        NaN       NaN         NaN        NaN      NaN   
191           NaN        NaN       NaN         NaN        NaN      NaN   
236           NaN        NaN       NaN         NaN        NaN      NaN   
...           ...        ...       ...         ...        ...      ...   
2999779       NaN        NaN       NaN         NaN        NaN      NaN   
2999865       NaN        NaN       NaN         NaN        NaN      NaN   
2999910       NaN        NaN       NaN         NaN        NaN      NaN   
2999943       NaN        NaN       NaN         NaN        NaN      NaN   
2999946       NaN        NaN       NaN         NaN        NaN      NaN   

         CRS_ARR_TIME  ARR_TIME  ARR_DELAY  CANCELLED CANCELLATION_CODE  \
14               2315       NaN        NaN        1.0                 D   
45               2241       NaN        NaN        1.0                 B   
80               1300       NaN        NaN        1.0                 D   
191              2024       NaN        NaN        1.0                 A   
236               955       NaN        NaN        1.0                 D   
...               ...       ...        ...        ...               ...   
2999779          2112       NaN        NaN        1.0                 C   
2999865          1719       NaN        NaN        1.0                 A   
2999910          1742       NaN        NaN        1.0                 D   
2999943          1655       NaN        NaN        1.0                 D   
2999946          1855       NaN        NaN        1.0                 C   

         DIVERTED  CRS_ELAPSED_TIME  ELAPSED_TIME  AIR_TIME  DISTANCE  \
14            0.0              80.0           NaN       NaN     308.0   
45            0.0             126.0           NaN       NaN     612.0   
80            0.0              65.0           NaN       NaN     214.0   
191           0.0             364.0           NaN       NaN    2288

so we see here that those 77615 rows have majority columns empty, so lets check if all of them were cancelled or diverted.

In [763]:
df[df['DEP_TIME'].isna()][['CANCELLED', 'DIVERTED']].value_counts()

CANCELLED  DIVERTED
1.0        0.0         77615
Name: count, dtype: int64

so all of them were cancelled! this gives alot of information. lets fill everything with 0, as all those times were 0.

In [764]:
null_rows.isna().sum()

FL_DATE                        0
AIRLINE                        0
AIRLINE_CODE                   0
DOT_CODE                       0
FL_NUMBER                      0
ORIGIN                         0
ORIGIN_CITY                    0
DEST                           0
DEST_CITY                      0
CRS_DEP_TIME                   0
DEP_TIME                   77615
DEP_DELAY                  77615
TAXI_OUT                   77615
WHEELS_OFF                 77615
WHEELS_ON                  77615
TAXI_IN                    77615
CRS_ARR_TIME                   0
ARR_TIME                   77615
ARR_DELAY                  77615
CANCELLED                      0
CANCELLATION_CODE              0
DIVERTED                       0
CRS_ELAPSED_TIME              13
ELAPSED_TIME               77615
AIR_TIME                   77615
DISTANCE                       0
DELAY_DUE_CARRIER          77615
DELAY_DUE_WEATHER          77615
DELAY_DUE_NAS              77615
DELAY_DUE_SECURITY         77615
DELAY_DUE_

so all of them are 77615 empty meaning all are empty. but 13 rows dont have `crs_elapsed_time` which can be computed using `crs_arr_time` - `crs_dep_time`. we will handle this later. lets handle the rest. 
### DEP_TIME, DEP_DELAY, TAXI_OUT, WHEELS_OFF, WHEELS_ON, TAXI_IN, ARR_TIME, ARR_DELAY, ELAPSED_TIME, AIR_TIME, DELAY_DUE_WEATHER, DELAY_DUE_NAS, DELAY_DUE_SECURITY, DELAY_DUE_LATE_AIRCRAFT

In [765]:
cols_to_fill = ['DEP_TIME', 'DEP_DELAY', 'TAXI_OUT', 'WHEELS_OFF', 'WHEELS_ON', 'TAXI_IN', 'ARR_TIME', 'ARR_DELAY', 'ELAPSED_TIME', 'AIR_TIME', 'DELAY_DUE_WEATHER', 'DELAY_DUE_NAS', 'DELAY_DUE_SECURITY', 'DELAY_DUE_LATE_AIRCRAFT']

In [766]:
# Loop over each column and fill NaNs with 0 in the original df using cancelled_only's index
for col in cols_to_fill:
    missing_idx = null_rows[null_rows[col].isna()].index
    df.loc[missing_idx, col] = 0

In [767]:
df['DEP_TIME'].isna().sum()

0

## DEP_DELAY

okay so we see that CRS_DEP_TIME and DEP_TIME is 0 null values that means we can compute DEP_DELAY null values using DEP_TIME - CRS_DEP_TIME to find the delay. however as both are in hhmm we will need to convert to minutes first 

In [768]:
df['DEP_DELAY'].isna().sum()

29

so we have 29 rows that are empty. lets first extract those 29 rows to work on

In [769]:
# Only fill DEP_DELAY where both times are present
mask = df[df['DEP_DELAY'].isna() & df['DEP_TIME'].notna() & df['CRS_DEP_TIME'].notna()]
mask

FL_DATE                AIRLINE AIRLINE_CODE  DOT_CODE  FL_NUMBER  \
86367   2021-08-12          Allegiant Air           G4     20368       2016   
135955  2021-06-21          Allegiant Air           G4     20368       2910   
148327  2021-07-24          Allegiant Air           G4     20368       1352   
270541  2021-07-01          Allegiant Air           G4     20368       2821   
387603  2021-10-31          Allegiant Air           G4     20368        229   
544736  2022-08-25          Allegiant Air           G4     20368       3302   
773927  2023-05-30  SkyWest Airlines Inc.           OO     20304       3053   
873432  2022-10-07          Allegiant Air           G4     20368        319   
1006686 2021-11-14          Allegiant Air           G4     20368       3662   
1223393 2021-06-13          Allegiant Air           G4     20368       2205   
1330580 2021-08-08          Allegiant Air           G4     20368        118   
1334049 2023-01-02          Allegiant Air           G4     20368       3068   
1515519 2020-01-17  SkyWest Airlines Inc.           OO     20304       3214   
1588106 2023-08-28          Allegiant Air           G4     20368       2265   
1729899 2021-10-11          Allegiant Air           G4     20368         74   
1744048 2021-08-01          Allegiant Air           G4     20368       2428   
2024315 2021-06-04          Allegiant Air           G4     20368       1251   
2037944 2020-01-23  SkyWest Airlines Inc.           OO     20304       3017   
2103457 2021-07-11          Allegiant Air           G4     20368       2203   
2125601 2020-01-17  SkyWest Airlines Inc.           OO     20304       5605   
2242217 2021-02-04          Allegiant Air           G4     20368        926   
2302250 2022-12-22          Allegiant Air           G4     20368       1800   
2609992 2023-07-28          Allegiant Air           G4     20368        183   
2612393 2020-09-13     Mesa Airlines Inc.           YV     20378       6130   
2672598 2020-01-24  SkyWest Airlines Inc.           OO     20304       3675   
2685999 2020-07-13          Allegiant Air           G4     20368        324   
2688640 2023-07-13          Allegiant Air           G4     20368       2809   
2710904 2022-04-16          Allegiant Air           G4     20368       2022   
2905853 2022-03-10          Allegiant Air           G4     20368       2266   

        ORIGIN                     ORIGIN_CITY DEST  \
86367      CLE                   Cleveland, OH  SRQ   
135955     SFB                     Sanford, FL  EYW   
148327     CVG                  Cincinnati, OH  VPS   
270541     LCK                    Columbus, OH  ORF   
387603     LAS                   Las Vegas, NV  FAT   
544736     SAN                   San Diego, CA  PVU   
773927     AMA                    Amarillo, TX  DFW   
873432     PIA                      Peoria, IL  DEN   
1006686    AZA                     Phoenix, AZ  PIT   
1223393    SFB                     Sanford, FL  BGR   
1330580    RNO                        Reno, NV  LAS   
1334049    VPS                  Valparaiso, FL  BLV   
1515519    ORD                     Chicago, IL  MCI   
1588106    AZA                     Phoenix, AZ  CVG   
1729899    MFR                     Medford, OR  LAS   
1744048    GRR                Grand Rapids, MI  LAX   
2024315    SFB                     Sanford, FL  AVL   
2037944    HPN                White Plains, NY  ORD   
2103457    SFB                     Sanford, FL  SYR   
2125601    ORD                     Chicago, IL  MCI   
2242217    TOL                      Toledo, OH  PIE   
2302250    BNA                   Nashville, TN  ABE   
2609992    ABE  Allentown/Bethlehem/Easton, PA  BNA   
2612393    RDU              Raleigh/Durham, NC  IAD   
2672598    MSP                 Minneapolis, MN  MDW   
2685999    BIL                    Billings, MT  LAX   
2688640    CAK                       Akron, OH  MYR   
2710904    SBN                  South Bend, IN  PIE   
2905853    SFB                     Sanford, F

In [770]:
mask.shape

(29, 31)

ok so we have extracted those rows. lets now apply the formula 

In [771]:
# Function to convert hhmm to total minutes since midnight
def hhmm_to_minutes(hhmm):
    hhmm = int(hhmm)
    hours = hhmm // 100
    minutes = hhmm % 100
    return hours * 60 + minutes

In [772]:
# Apply the conversion and calculate DEP_DELAY for masked rows
df.loc[mask.index, 'DEP_DELAY'] = (
    mask['DEP_TIME'].apply(hhmm_to_minutes) -
    mask['CRS_DEP_TIME'].apply(hhmm_to_minutes)
)

df['DEP_DELAY'].isna().sum()

0

In [773]:
df['DEP_DELAY'].value_counts()

DEP_DELAY
-5.0       240153
-4.0       223241
 0.0       219561
-3.0       212650
-2.0       191413
            ...  
 1349.0         1
 1351.0         1
 1071.0         1
 2206.0         1
 1647.0         1
Name: count, Length: 1515, dtype: int64

## TAXI_OUT
so `taxi_out` is the time from gate pushback to takeoff.
so here lets first see how many rows are empty

In [774]:
df['TAXI_OUT'].isnull().sum()

1191

so now 1191 is not a very large value out of 3 million. we have one option to drop those rows. but lets think of something better - lets check those `taxi_out` rows, are they all cancelled or diverted flights - then `taxi_out` will be null. lets extract the cancelled and diverted status of those rows.

In [775]:
df[df['TAXI_OUT'].isna()][['CANCELLED', 'DIVERTED']].value_counts()

CANCELLED  DIVERTED
1.0        0.0         1191
Name: count, dtype: int64

so we see that in all those 1191 rows, those flights are CANCELLED. thus the plane has never left the origin - thus there is no `taxi_out`. respectively, there won't be any `wheels_off`, `wheels_on`, `arr_time`, `taxi_in`, `air_time`, `elapsed_time` either. so first lets extract those `taxi_out` rows, and then lets extract ALL the cancelled rows and see what is `taxi_out` is filled? lets fill that with it exactly to maintain consistency.

In [776]:
# Extract rows with CANCELLED = 1 and DIVERTED = 0
cancelled_only = df[(df['CANCELLED'] == 1) & (df['DIVERTED'] == 0)]
cancelled_only.shape

(79140, 31)

In [777]:
cancelled_only['TAXI_OUT'].isna().sum()

1191

In [778]:
cancelled_only['TAXI_OUT'].notna().sum()

77949

so there does exist 77949 rows that have `taxi_out` values, lets check their values

In [779]:
cancelled_only['TAXI_OUT'].value_counts()

TAXI_OUT
0.0      77615
13.0        24
15.0        24
12.0        23
20.0        21
18.0        20
14.0        18
16.0        17
17.0        14
10.0        13
22.0        12
11.0        12
9.0         11
24.0        11
21.0        10
19.0        10
29.0         9
27.0         7
25.0         7
23.0         6
30.0         5
8.0          4
41.0         4
32.0         4
31.0         4
39.0         4
26.0         4
40.0         3
28.0         2
87.0         2
71.0         2
46.0         2
60.0         2
38.0         2
52.0         2
36.0         2
34.0         2
33.0         2
95.0         1
47.0         1
55.0         1
37.0         1
54.0         1
94.0         1
67.0         1
56.0         1
42.0         1
7.0          1
123.0        1
64.0         1
43.0         1
Name: count, dtype: int64

so this is weird. even if the flight is cancelled, there is a taxi_out. this might mean the flight was ready to take off, but was cancelled before the wheels started moving. lets see if we have a `wheels_off` for these cancelled rows

In [780]:
cancelled_only['WHEELS_OFF'].isna().sum()

1191

In [781]:
cancelled_only['WHEELS_OFF'].value_counts()

WHEELS_OFF
0.0       77615
2019.0        3
2254.0        3
1557.0        3
2318.0        3
          ...  
2129.0        1
1322.0        1
2140.0        1
1005.0        1
1504.0        1
Name: count, Length: 285, dtype: int64

look how the majority values of CANCLEDD FLIGHTS have 0. this means that we fill cancelled flights with 0      
okay so this is great - our suspicions have been confirmed. the plane had taken off, but was cancelled before reaching the destination, maybe due to bad weather or a technical fault. lets extract those cancelled reasons too

In [782]:
cancelled_only['CANCELLATION_CODE'].value_counts()

CANCELLATION_CODE
B    28772
D    24417
A    19476
C     6475
Name: count, dtype: Int64

ah so majority was due to weather. so lets fill with some value to show that there is no value. lets say 0. but before we do that, lets extract all the null values in these cancelled rows and get their counts so that we can fill with 0

In [783]:
cancelled_nulls = cancelled_only[cancelled_only['TAXI_OUT'].isna()]
cancelled_nulls.isna().sum()

FL_DATE                       0
AIRLINE                       0
AIRLINE_CODE                  0
DOT_CODE                      0
FL_NUMBER                     0
ORIGIN                        0
ORIGIN_CITY                   0
DEST                          0
DEST_CITY                     0
CRS_DEP_TIME                  0
DEP_TIME                      0
DEP_DELAY                     0
TAXI_OUT                   1191
WHEELS_OFF                 1191
WHEELS_ON                  1191
TAXI_IN                    1191
CRS_ARR_TIME                  0
ARR_TIME                   1191
ARR_DELAY                  1191
CANCELLED                     0
CANCELLATION_CODE             0
DIVERTED                      0
CRS_ELAPSED_TIME              1
ELAPSED_TIME               1191
AIR_TIME                   1191
DISTANCE                      0
DELAY_DUE_CARRIER          1191
DELAY_DUE_WEATHER          1191
DELAY_DUE_NAS              1191
DELAY_DUE_SECURITY         1191
DELAY_DUE_LATE_AIRCRAFT    1191
dtype: i

now lets fill all the empty cells with 0
### TAXI_OUT, WHEELS_OFF, WHEELS_ON, TAXI_IN, ARR_TIME, ARR_DELAY, ELAPSED_TIME, AIR_TIME, DELAY_DUE_CARRIER, DELAY_DUE_WEATHER, DELAY_DUE_NAS, DELAY_DUE_SECURITY, DELAY_DUE_LATE_AIRCRAFT

In [784]:
cols_to_fill = ['TAXI_OUT', 'WHEELS_OFF', 'WHEELS_ON', 'TAXI_IN', 'ARR_TIME', 'ARR_DELAY', 'ELAPSED_TIME', 'AIR_TIME', 'DELAY_DUE_CARRIER', 'DELAY_DUE_WEATHER', 'DELAY_DUE_NAS', 'DELAY_DUE_SECURITY', 'DELAY_DUE_LATE_AIRCRAFT']

In [785]:
# Loop over each column and fill NaNs with 0 in the original df using cancelled_nulls's index
for col in cols_to_fill:
    missing_idx = cancelled_nulls[cancelled_nulls[col].isna()].index
    df.loc[missing_idx, col] = 0

In [786]:
df['TAXI_OUT'].isna().sum()

0

## WHEELS_OFF
none are empty

In [787]:
df['WHEELS_OFF'].isna().sum()

0

## WHEELS_ON

In [788]:
df['WHEELS_ON'].isna().sum()

1138

so we found some null values. lets see the status of all these null values

In [789]:
df[df['WHEELS_ON'].isna()][['CANCELLED', 'DIVERTED']].value_counts()

CANCELLED  DIVERTED
0.0        1.0         802
1.0        0.0         334
0.0        0.0           2
Name: count, dtype: int64

so 802 flights are diverted menaing that did not land at the intended destination. It was rerouted in the air due to weather, technical issues, airport congestion, etc. 2 flights were neither cancelled nor diverted so we will deal with these first.

In [790]:
rows_null = df[df['WHEELS_ON'].isna()]
data_missing_rows = rows_null[(rows_null['DIVERTED'] == 0.0) & (rows_null['CANCELLED'] == 0.0)]
data_missing_rows

FL_DATE                 AIRLINE AIRLINE_CODE  DOT_CODE  FL_NUMBER  \
918330  2023-05-21        Republic Airline           YX     20452       5612   
1302862 2022-12-21  Frontier Airlines Inc.           F9     20436        777   

        ORIGIN   ORIGIN_CITY DEST      DEST_CITY  CRS_DEP_TIME  DEP_TIME  \
918330     LGA  New York, NY  BNA  Nashville, TN          1605    1642.0   
1302862    DEN    Denver, CO  LAS  Las Vegas, NV          1956    1951.0   

         DEP_DELAY  TAXI_OUT  WHEELS_OFF  WHEELS_ON  TAXI_IN  CRS_ARR_TIME  \
918330        37.0      36.0      1718.0        NaN      NaN          1746   
1302862       -5.0      81.0      2112.0        NaN      NaN          2100   

         ARR_TIME  ARR_DELAY  CANCELLED CANCELLATION_CODE  DIVERTED  \
918330        NaN        NaN        0.0              <NA>       0.0   
1302862       NaN        NaN        0.0              <NA>       0.0   

         CRS_ELAPSED_TIME  ELAPSED_TIME  AIR_TIME  DISTANCE  \
918330              161.0           NaN       NaN     764.0   
1302862             124.0           NaN       NaN     628.0   

         DELAY_DUE_CARRIER  DELAY_DUE_WEATHER  DELAY_DUE_NAS  \
918330                 NaN                NaN            NaN   
1302862                NaN                NaN            NaN   

         DELAY_DUE_SECURITY  DELAY_DUE_LATE_AIRCRAFT  
918330                  NaN                      NaN  
1302862                 NaN                      NaN

so these 2 rows, they did take off, they were not diverted, and were not cancelled - yet we dont have any data of them landing, no `wheels_on`, `taxi_in`, `arr_time`. thus this is a data missing entry and we cannot deal with it. it is better to drop. 

In [791]:
df = df.drop(index=data_missing_rows.index)

so lets extract those 802 flights first and fill all their null values with 0 like we did for all the cancelled flights. we will also do this for the cancelled flights

In [792]:
rows_null.shape

(1138, 31)

In [793]:
rows_null

FL_DATE                 AIRLINE AIRLINE_CODE  DOT_CODE  FL_NUMBER  \
2523    2020-10-27  Southwest Airlines Co.           WN     19393       1227   
2870    2020-01-17   United Air Lines Inc.           UA     19977       1897   
2875    2019-05-29  American Airlines Inc.           AA     19805       2331   
6368    2019-06-08   SkyWest Airlines Inc.           OO     20304       5345   
16778   2020-12-19    Alaska Airlines Inc.           AS     19930         60   
...            ...                     ...          ...       ...        ...   
2976600 2022-03-17  Southwest Airlines Co.           WN     19393       2531   
2979451 2022-01-03    Alaska Airlines Inc.           AS     19930         65   
2981404 2022-05-22   United Air Lines Inc.           UA     19977       2208   
2984141 2019-04-10   SkyWest Airlines Inc.           OO     20304       5111   
2995212 2023-02-26   SkyWest Airlines Inc.           OO     20304       5086   

        ORIGIN           ORIGIN_CITY DEST              DEST_CITY  \
2523       DEN            Denver, CO  ICT            Wichita, KS   
2870       LAX       Los Angeles, CA  ORD            Chicago, IL   
2875       CLE         Cleveland, OH  DFW  Dallas/Fort Worth, TX   
6368       VPS        Valparaiso, FL  DEN             Denver, CO   
16778      JNU            Juneau, AK  KTN          Ketchikan, AK   
...        ...                   ...  ...                    ...   
2976600    STL         St. Louis, MO  LGA           New York, NY   
2979451    PSG        Petersburg, AK  JNU             Juneau, AK   
2981404    STT  Charlotte Amalie, VI  SJU           San Juan, PR   
2984141    DEN            Denver, CO  EAR            Kearney, NE   
2995212    DEN            Denver, CO  LBL            Liberal, KS   

         CRS_DEP_TIME  DEP_TIME  DEP_DELAY  TAXI_OUT  WHEELS_OFF  WHEELS_ON  \
2523             1220    1234.0       14.0       9.0      1243.0        NaN   
2870             1403    1447.0       44.0      13.0      1500.0        NaN   
2875             1145    1151.0        6.0      17.0      1208.0        NaN   
6368             1535    1547.0       12.0      25.0      1612.0        NaN   
16778             720     741.0       21.0       8.0       749.0        NaN   
...               ...       ...        ...       ...         ...        ...   
2976600          1350    1548.0      118.0      22.0      1610.0        NaN   
2979451          1145    1257.0       72.0      26.0      1323.0        NaN   
2981404          1316    1354.0       38.0      15.0      1409.0        NaN   
2984141          1325    1312.0      -13.0      34.0      1346.0        NaN   
2995212          1449    1446.0       -3.0      18.0      1504.0        NaN   

         TAXI_IN  CRS_ARR_TIME  ARR_TIME  ARR_DELAY  CANCELLED  \
2523         NaN          1435       NaN        NaN        0.0   
2870         NaN          2009       NaN        NaN        1.0   
2875         NaN          1338       NaN        NaN        0.0   
6368         NaN          1745       NaN        NaN        0.0   
16778        NaN           828       NaN        NaN        0.0   
...          ...           ...       ...        ...        ...   
2976600      NaN          1710       NaN        NaN        0.0   
2979451      NaN          1231       NaN        NaN        0.0   
2981404      NaN          1353       NaN        NaN        0.0   
2984141      NaN          1543       NaN        NaN        0.0   
2995212      NaN          1658       NaN        NaN        1.0   

        CANCELLATION_CODE  DIVERTED  CRS_ELAPSED_TIME  ELAPSED_TIME  AIR_TIME  \
2523                 <NA>       1.0              75.0           NaN       NaN   
2870                    B       0.0             246.0           NaN       NaN   
2875                 <NA>       1.0             173.0           NaN       NaN   
6368                 <NA>       1.0             190.0           NaN       NaN   
16778                <NA>       1.0              68.0           NaN       NaN   
...  

first lets do it for the diverted flights

In [794]:
# Extract rows with CANCELLED = 0 and DIVERTED = 1
diverted_only = rows_null[(rows_null['CANCELLED'] == 0) & (rows_null['DIVERTED'] == 1)]
diverted_only.shape

(802, 31)

In [795]:
diverted_only.isna().sum()

FL_DATE                      0
AIRLINE                      0
AIRLINE_CODE                 0
DOT_CODE                     0
FL_NUMBER                    0
ORIGIN                       0
ORIGIN_CITY                  0
DEST                         0
DEST_CITY                    0
CRS_DEP_TIME                 0
DEP_TIME                     0
DEP_DELAY                    0
TAXI_OUT                     0
WHEELS_OFF                   0
WHEELS_ON                  802
TAXI_IN                    802
CRS_ARR_TIME                 0
ARR_TIME                   800
ARR_DELAY                  802
CANCELLED                    0
CANCELLATION_CODE          802
DIVERTED                     0
CRS_ELAPSED_TIME             0
ELAPSED_TIME               802
AIR_TIME                   802
DISTANCE                     0
DELAY_DUE_CARRIER          802
DELAY_DUE_WEATHER          802
DELAY_DUE_NAS              802
DELAY_DUE_SECURITY         802
DELAY_DUE_LATE_AIRCRAFT    802
dtype: int64

so we spotted something new: `arr_time` does have 2 entries which the others do not have. lets work on those and extract those rows and fill them accordingly.

In [796]:
arr_time_full = diverted_only[diverted_only['ARR_TIME'].notna()]
arr_time_full

FL_DATE                 AIRLINE AIRLINE_CODE  DOT_CODE  FL_NUMBER  \
1296624 2021-10-20  Southwest Airlines Co.           WN     19393       1284   
1877623 2022-01-02   SkyWest Airlines Inc.           OO     20304       3711   

        ORIGIN      ORIGIN_CITY DEST        DEST_CITY  CRS_DEP_TIME  DEP_TIME  \
1296624    LAX  Los Angeles, CA  BWI    Baltimore, MD           755     759.0   
1877623    SAT  San Antonio, TX  LAX  Los Angeles, CA           608     602.0   

         DEP_DELAY  TAXI_OUT  WHEELS_OFF  WHEELS_ON  TAXI_IN  CRS_ARR_TIME  \
1296624        4.0      14.0       813.0        NaN      NaN          1550   
1877623       -6.0      14.0       616.0        NaN      NaN           725   

         ARR_TIME  ARR_DELAY  CANCELLED CANCELLATION_CODE  DIVERTED  \
1296624    1750.0        NaN        0.0              <NA>       1.0   
1877623     904.0        NaN        0.0              <NA>       1.0   

         CRS_ELAPSED_TIME  ELAPSED_TIME  AIR_TIME  DISTANCE  \
1296624             295.0           NaN       NaN    2329.0   
1877623             197.0           NaN       NaN    1211.0   

         DELAY_DUE_CARRIER  DELAY_DUE_WEATHER  DELAY_DUE_NAS  \
1296624                NaN                NaN            NaN   
1877623                NaN                NaN            NaN   

         DELAY_DUE_SECURITY  DELAY_DUE_LATE_AIRCRAFT  
1296624                 NaN                      NaN  
1877623                 NaN                      NaN

In [797]:
arr_time_full.isna().sum()

FL_DATE                    0
AIRLINE                    0
AIRLINE_CODE               0
DOT_CODE                   0
FL_NUMBER                  0
ORIGIN                     0
ORIGIN_CITY                0
DEST                       0
DEST_CITY                  0
CRS_DEP_TIME               0
DEP_TIME                   0
DEP_DELAY                  0
TAXI_OUT                   0
WHEELS_OFF                 0
WHEELS_ON                  2
TAXI_IN                    2
CRS_ARR_TIME               0
ARR_TIME                   0
ARR_DELAY                  2
CANCELLED                  0
CANCELLATION_CODE          2
DIVERTED                   0
CRS_ELAPSED_TIME           0
ELAPSED_TIME               2
AIR_TIME                   2
DISTANCE                   0
DELAY_DUE_CARRIER          2
DELAY_DUE_WEATHER          2
DELAY_DUE_NAS              2
DELAY_DUE_SECURITY         2
DELAY_DUE_LATE_AIRCRAFT    2
dtype: int64

now for this specific case we can't fill all the rows with 0, there does exist `arr_time` so thus `elapsed_time` can be computed, `air_time` cannot as we dont have `wheels_on`. so lets just make `wheels_on`, `taxi_in`, `air_time` with 0s, the rest can be filled later on.

In [798]:
cols_to_fill = ['WHEELS_ON', 'TAXI_IN', 'AIR_TIME']
for col in cols_to_fill:
    missing_idx = arr_time_full[arr_time_full[col].isna()].index
    df.loc[missing_idx, col] = 0

now lets remove these two columns from `diverted_only` dataframe

In [799]:
diverted_only = diverted_only[diverted_only['ARR_TIME'].isna()]
diverted_only.isna().sum()

FL_DATE                      0
AIRLINE                      0
AIRLINE_CODE                 0
DOT_CODE                     0
FL_NUMBER                    0
ORIGIN                       0
ORIGIN_CITY                  0
DEST                         0
DEST_CITY                    0
CRS_DEP_TIME                 0
DEP_TIME                     0
DEP_DELAY                    0
TAXI_OUT                     0
WHEELS_OFF                   0
WHEELS_ON                  800
TAXI_IN                    800
CRS_ARR_TIME                 0
ARR_TIME                   800
ARR_DELAY                  800
CANCELLED                    0
CANCELLATION_CODE          800
DIVERTED                     0
CRS_ELAPSED_TIME             0
ELAPSED_TIME               800
AIR_TIME                   800
DISTANCE                     0
DELAY_DUE_CARRIER          800
DELAY_DUE_WEATHER          800
DELAY_DUE_NAS              800
DELAY_DUE_SECURITY         800
DELAY_DUE_LATE_AIRCRAFT    800
dtype: int64

so lets fill all the empty values with 0
### WHEELS_ON, TAXI_IN, ARR_TIME, ARR_DELAY, CANCELLATION_CODE, ELAPSED_TIME, AIR_TIME, DELAY_DUE_CARRIER, DELAY_DUE_WEATHER, DELAY_DUE_NAS, DELAY_DUE_SECURITY, DELAY_DUE_LATE_AIRCRAFT

In [800]:
cols_to_fill = ['WHEELS_ON', 'TAXI_IN', 'ARR_TIME', 'ARR_DELAY', 'ELAPSED_TIME', 'AIR_TIME', 'DELAY_DUE_CARRIER', 'DELAY_DUE_WEATHER', 'DELAY_DUE_NAS', 'DELAY_DUE_SECURITY', 'DELAY_DUE_LATE_AIRCRAFT']

In [801]:
# Loop over each column and fill NaNs with 0 in the original df using diverted_only's index
for col in cols_to_fill:
    missing_idx = diverted_only[diverted_only[col].isna()].index
    df.loc[missing_idx, col] = 0

In [802]:
df['WHEELS_ON'].isna().sum()

334

now lets do it for the cancelled flights

In [803]:
# Extract rows with CANCELLED = 0 and DIVERTED = 1
cancelled_only = rows_null[(rows_null['CANCELLED'] == 1) & (rows_null['DIVERTED'] == 0)]
cancelled_only.shape

(334, 31)

In [804]:
cancelled_only.isna().sum()

FL_DATE                      0
AIRLINE                      0
AIRLINE_CODE                 0
DOT_CODE                     0
FL_NUMBER                    0
ORIGIN                       0
ORIGIN_CITY                  0
DEST                         0
DEST_CITY                    0
CRS_DEP_TIME                 0
DEP_TIME                     0
DEP_DELAY                    0
TAXI_OUT                     0
WHEELS_OFF                   0
WHEELS_ON                  334
TAXI_IN                    334
CRS_ARR_TIME                 0
ARR_TIME                   334
ARR_DELAY                  334
CANCELLED                    0
CANCELLATION_CODE            0
DIVERTED                     0
CRS_ELAPSED_TIME             0
ELAPSED_TIME               334
AIR_TIME                   334
DISTANCE                     0
DELAY_DUE_CARRIER          334
DELAY_DUE_WEATHER          334
DELAY_DUE_NAS              334
DELAY_DUE_SECURITY         334
DELAY_DUE_LATE_AIRCRAFT    334
dtype: int64

### WHEELS_ON, TAXI_IN, ARR_TIME, ARR_DELAY, ELAPSED_TIME, AIR_TIME, DELAY_DUE_CARRIER, DELAY_DUE_WEATHER, DELAY_DUE_NAS, DELAY_DUE_SECURITY, DELAY_DUE_LATE_AIRCRAFT

In [805]:
cols_to_fill = ['WHEELS_ON', 'TAXI_IN', 'ARR_TIME', 'ARR_DELAY', 'ELAPSED_TIME', 'AIR_TIME', 'DELAY_DUE_CARRIER', 'DELAY_DUE_WEATHER', 'DELAY_DUE_NAS', 'DELAY_DUE_SECURITY', 'DELAY_DUE_LATE_AIRCRAFT']

In [806]:
# Loop over each column and fill NaNs with 0 in the original df using cancelled_only's index
for col in cols_to_fill:
    missing_idx = cancelled_only[cancelled_only[col].isna()].index
    df.loc[missing_idx, col] = 0

In [807]:
df['WHEELS_ON'].isna().sum()

0

## TAXI_IN
none are empty

In [808]:
df['TAXI_IN'].isna().sum()

0

## ARR_TIME
none are empty

In [809]:
df['ARR_TIME'].isna().sum()

0

## ARR_DELAY
so as we discussed about `dep_delay` we said that it was the difference between the scheduled departure time and actual departure time. here it is the same thing - `arr_delay` is the difference between the scheduled arrival time and the actual arrival time. lets now do the same implementation we did for `dep_delay`.

In [810]:
df['ARR_DELAY'].isnull().sum()

6256

In [811]:
mask = df[df['ARR_DELAY'].isna() & df['ARR_TIME'].notna() & df['CRS_ARR_TIME'].notna()]
mask.shape

(6256, 31)

In [812]:
df.loc[mask.index, 'ARR_DELAY'] = (
    mask['ARR_TIME'].apply(hhmm_to_minutes) -
    mask['CRS_ARR_TIME'].apply(hhmm_to_minutes)
)

In [813]:
df['ARR_DELAY'].isnull().sum()

0

## CANCELLATION_CODE
so we can see that the cancellation_code empty values are BECAUSE THAT FLIGHT WAS NOT CANCELLED. AS IT IS A STRING type we can just fill with a string of "not cancelled"

In [814]:
df['CANCELLATION_CODE'].isna().sum()

2920858

In [815]:
df['CANCELLATION_CODE'].notna().sum()

79140

In [816]:
df['CANCELLATION_CODE'].value_counts()

CANCELLATION_CODE
B    28772
D    24417
A    19476
C     6475
Name: count, dtype: Int64

so here we see that this is not interpretable. we searched up and dug up the internet to get to know that from the U.S. Department of Transportation (DOT) flight data standards, here is what each cancellation code means:
- A = Carrier — airline-related (crew, aircraft, etc.)
- B = Weather — adverse weather at origin/destination or en route
- C = NAS — National Airspace System (ATC, delays, airport congestion)
- D = Security — e.g., TSA delays, threats, evacuations     

so instead of leaving it just like this, lets make it more visualizeable.

along with mapping each one, lets replace all null values with "Not Cancelled"

In [817]:
cancel_map = {
    'A': 'Carrier',
    'B': 'Weather',
    'C': 'Airspace System (NAS)',
    'D': 'Security',
    'Not Cancelled': 'Not Cancelled'
}

df['CANCELLATION_CODE'] = df['CANCELLATION_CODE'].fillna('Not Cancelled').map(cancel_map)
df = to_string(df, 'CANCELLATION_CODE')

Column Name: 'CANCELLATION_CODE', Type: string


In [818]:
df['CANCELLATION_CODE'].value_counts()

CANCELLATION_CODE
Not Cancelled            2920858
Weather                    28772
Security                   24417
Carrier                    19476
Airspace System (NAS)       6475
Name: count, dtype: Int64

In [819]:
df['CANCELLATION_CODE'].isna().sum()

0

## CRS_ELAPSED_TIME
so we know we can compute this as the difference between the `CRS_DEP_TIME` and the `CRS_ARR_TIME`. same as `DEP_DELAY` we first compute the hhmm accordingly and then compute differnece

In [820]:
df['CRS_ELAPSED_TIME'].isnull().sum()

14

In [821]:
mask = df[df['CRS_ELAPSED_TIME'].isna() & df['CRS_ARR_TIME'].notna() & df['CRS_DEP_TIME'].notna()]
mask.shape

(14, 31)

yep so we have extracted all the relevant rows. lets apply it on them.

In [822]:
scheduled_elapsed = (
    mask['CRS_ARR_TIME'].apply(hhmm_to_minutes) -
    mask['CRS_DEP_TIME'].apply(hhmm_to_minutes)
)

scheduled_elapsed = scheduled_elapsed.apply(lambda x: x + 1440 if x < 0 else x)

df.loc[mask.index, 'CRS_ELAPSED_TIME'] = scheduled_elapsed

In [823]:
df['CRS_ELAPSED_TIME'].isnull().sum()

0

## ELAPSED_TIME
as discussed in dep_delay, elapsed_time is the time of the flight. its the actual elapsed time so same as `crs_elapsed_time`, its just `arr_time` - `dep_time`. so lets do the same thing as above.

In [824]:
df['ELAPSED_TIME'].isnull().sum()

6256

In [825]:
mask = df[df['ELAPSED_TIME'].isna() & df['ARR_TIME'].notna() & df['DEP_TIME'].notna()]
mask.shape

(6256, 31)

In [826]:
def safe_hhmm_to_minutes(hhmm):
    try:
        hhmm = int(hhmm)
        return (hhmm // 100) * 60 + (hhmm % 100)
    except:
        return None

In [827]:
elapsed_minutes = (
    mask['ARR_TIME'].apply(safe_hhmm_to_minutes) -
    mask['DEP_TIME'].apply(safe_hhmm_to_minutes)
)

# Adjust for overnight flights
elapsed_minutes = elapsed_minutes.apply(lambda x: x + 1440 if x < 0 else x)

df.loc[mask.index, 'ELAPSED_TIME'] = elapsed_minutes

In [828]:
df['ELAPSED_TIME'].isnull().sum()

0

## AIR_TIME
so now this is same as `crs_elapsed_time` and `elapsed_time`. `air_time` is just `wheels_on` and `wheels_off` difference. so lets do the same implementation

In [829]:
df['AIR_TIME'].isnull().sum()

6254

In [830]:
mask = df[df['AIR_TIME'].isna() & df['WHEELS_ON'].notna() & df['WHEELS_OFF'].notna()]
mask.shape

(6254, 31)

In [831]:
air_minutes = (
    mask['WHEELS_ON'].apply(safe_hhmm_to_minutes) -
    mask['WHEELS_OFF'].apply(safe_hhmm_to_minutes)
)

air_minutes = air_minutes.apply(lambda x: x + 1440 if x < 0 else x)

df.loc[mask.index, 'AIR_TIME'] = air_minutes

In [832]:
df['AIR_TIME'].isnull().sum()

0

## DELAY_DUE_CARRIER, DELAY_DUE_WEATHER, DELAY_DUE_NAS, DELAY_DUE_SECURITY, DELAY_DUE_LATE_AIRCRAFT
now every column here has one thing in common: same missing values. this also means that whenever there was NOT A DELAY, then they were left blank. so it is better to fill them with 0s to show that there was NO DELAY. if i fill them with -1, it will hinder the mean/median/average, and saying -1minutes of delay does not make sense, wheras 0 minutes of delay does.

In [833]:
df['DELAY_DUE_CARRIER'].isnull().sum()

2463810

In [834]:
df['DELAY_DUE_CARRIER'].value_counts()

DELAY_DUE_CARRIER
0.0       239237
15.0        9102
1.0         9073
2.0         8941
3.0         8670
           ...  
1117.0         1
2565.0         1
998.0          1
1192.0         1
1859.0         1
Name: count, Length: 1291, dtype: int64

so we see from the value_counts that the majority is 0s which means there is no delay of this cause. so we will fill all NANs wth 0s

In [835]:
delay_columns = ['DELAY_DUE_CARRIER', 'DELAY_DUE_WEATHER', 'DELAY_DUE_NAS', 'DELAY_DUE_SECURITY', 'DELAY_DUE_LATE_AIRCRAFT']
for col in delay_columns:
    df[col] = df[col].fillna(0.0)
    
df['DELAY_DUE_CARRIER'].isnull().sum()

0

## New columns
here we analyse if we need new columns in integers or not for example of destination city etc so that we can see common ones

### FL_DATE
here we add 4 new columns for year, quarter, month, day to make visualisation easier

In [836]:
# for FL_DATE, add 4 new columns which have the year, quarter, month, date for better preprocessing and display
df['year'] = df['FL_DATE'].dt.year
df['quarter'] = df['FL_DATE'].dt.quarter
df['month'] = df['FL_DATE'].dt.strftime('%b')  # 3-letter month name
df = to_string(df, 'month')
df['day'] = df['FL_DATE'].dt.day

df

Column Name: 'month', Type: string


FL_DATE                 AIRLINE AIRLINE_CODE  DOT_CODE  FL_NUMBER  \
0       2019-01-09   United Air Lines Inc.           UA     19977       1562   
1       2022-11-19    Delta Air Lines Inc.           DL     19790       1149   
2       2022-07-22   United Air Lines Inc.           UA     19977        459   
3       2023-03-06    Delta Air Lines Inc.           DL     19790       2295   
4       2020-02-23        Spirit Air Lines           NK     20416        407   
...            ...                     ...          ...       ...        ...   
2999995 2022-11-13  American Airlines Inc.           AA     19805       1522   
2999996 2022-11-02  American Airlines Inc.           AA     19805       1535   
2999997 2022-09-11    Delta Air Lines Inc.           DL     19790       2745   
2999998 2019-11-13        Republic Airline           YX     20452       6134   
2999999 2019-06-15  Southwest Airlines Co.           WN     19393       2823   

        ORIGIN          ORIGIN_CITY DEST              DEST_CITY  CRS_DEP_TIME  \
0          FLL  Fort Lauderdale, FL  EWR             Newark, NJ          1155   
1          MSP      Minneapolis, MN  SEA            Seattle, WA          2120   
2          DEN           Denver, CO  MSP        Minneapolis, MN           954   
3          MSP      Minneapolis, MN  SFO      San Francisco, CA          1609   
4          MCO          Orlando, FL  DFW  Dallas/Fort Worth, TX          1840   
...        ...                  ...  ...                    ...           ...   
2999995    JAX     Jacksonville, FL  CLT          Charlotte, NC          1742   
2999996    ORD          Chicago, IL  AUS             Austin, TX          1300   
2999997    HSV       Huntsville, AL  ATL            Atlanta, GA           534   
2999998    BOS           Boston, MA  LGA           New York, NY          1600   
2999999    LGB       Long Beach, CA  SJC           San Jose, CA           730   

         DEP_TIME  DEP_DELAY  TAXI_OUT  WHEELS_OFF  WHEELS_ON  TAXI_IN  \
0          1151.0       -4.0      19.0      1210.0     1443.0      4.0   
1          2114.0       -6.0       9.0      2123.0     2232.0     38.0   
2          1000.0        6.0      20.0      1020.0     1247.0      5.0   
3          1608.0       -1.0      27.0      1635.0     1844.0      9.0   
4          1838.0       -2.0      15.0      1853.0     2026.0     14.0   
...           ...        ...       ...         ...        ...      ...   
2999995    1740.0       -2.0      10.0      1750.0     1845.0      6.0   
2999996    1254.0       -6.0      10.0      1304.0     1514.0      5.0   
2999997     615.0       41.0      16.0       631.0      759.0      6.0   
2999998    1555.0       -5.0      19.0      1614.0     1704.0      8.0   
2999999     727.0       -3.0       9.0       736.0      828.0      2.0   

         CRS_ARR_TIME  ARR_TIME  ARR_DELAY  CANCELLED CANCELLATION_CODE  \
0                1501    1447.0      -14.0        0.0     Not Cancelled   
1                2315    2310.0       -5.0        0.0     Not Cancelled   
2                1252    1252.0        0.0        0.0     Not Cancelled   
3                1829    1853.0       24.0        0.0     Not Cancelled   
4                2041    2040.0       -1.0        0.0     Not Cancelled   
...               ...       ...        ...        ...               ...   
2999995          1907    1851.0      -16.0        0.0     Not Cancelled   
2999996          1556    1519.0      -37.0        0.0     Not Cancelled   
2999997           729     805.0       36.0        0.0     Not Cancelled   
2999998          1728    1712.0      -16.0        0.0     Not Cancelled   
2999999           840     830.0      -10.0        0.0     Not Cancelled   

         DIVERTED  CRS_ELAPSED_TIME  ELAPSED_TIME  AIR_TIME  DISTANCE  \
0             0.0             186.0         176.0     153.0    1065.0   
1             0.0             235.0         236.0     189.0    1399.0   
2             0.0             118.0         112.0      87.0     680.

In [837]:
# Move the new columns after FL_DATE
# Get all column names
cols = list(df.columns)

# Get index of FL_DATE
fl_date_index = cols.index('FL_DATE')

# Remove the new columns from the end
for col in ['year', 'quarter', 'month', 'day']:
    cols.remove(col)

# Insert new columns right after FL_DATE
for i, col in enumerate(['year', 'quarter', 'month', 'day']):
    cols.insert(fl_date_index + 1 + i, col)

# Reorder the DataFrame
df = df[cols]

### Categorical Columns
so the CANCELLED column and DIVERTED column are according to `dictionary.html` binary values of 0 1, so to improve visualisation in BI, we can convert them to categorical values of 'No' and 'Yes' respectively.

In [838]:
df['CANCELLED'].value_counts()

CANCELLED
0.0    2920858
1.0      79140
Name: count, dtype: int64

In [839]:
df['DIVERTED'].value_counts()

DIVERTED
0.0    2992942
1.0       7056
Name: count, dtype: int64

so we can convert CANCELLED and DIVERTED to integers, and then add columns for categorical

In [840]:
df = to_int(df, 'CANCELLED')
df = to_int(df, 'DIVERTED')

Column Name: 'CANCELLED', Type: int64
Column Name: 'DIVERTED', Type: int64


In [841]:
df['DIVERTED'].value_counts()

DIVERTED
0    2992942
1       7056
Name: count, dtype: int64

In [842]:
df['CANCELLED'].value_counts()

CANCELLED
0    2920858
1      79140
Name: count, dtype: int64

In [843]:
# Mapping for a new column
mapping = {
    0: 'No', 
    1: 'Yes'
}

# Create a new column 
df['cancelled_c'] = df['CANCELLED'].map(mapping)
df = to_string(df, 'cancelled_c')

df['diverted_c'] = df['DIVERTED'].map(mapping)
df = to_string(df, 'diverted_c')

Column Name: 'cancelled_c', Type: string
Column Name: 'diverted_c', Type: string


In [844]:
df['cancelled_c'].value_counts()

cancelled_c
No     2920858
Yes      79140
Name: count, dtype: Int64

In [845]:
df['diverted_c'].value_counts()

diverted_c
No     2992942
Yes       7056
Name: count, dtype: Int64

## Changing Column names
to improve understandability

In [846]:
df.rename(columns={'CRS_DEP_TIME': 'SCHEDULED_DEP_TIME'}, inplace=True)
df.rename(columns={'DEP_TIME': 'ACTUAL_DEP_TIME'}, inplace=True)
df.rename(columns={'CRS_ARR_TIME': 'SCHEDULED_ARR_TIME'}, inplace=True)
df.rename(columns={'ARR_TIME': 'ACTUAL_ARR_TIME'}, inplace=True)
df.rename(columns={'CRS_ELAPSED_TIME': 'SCHEDULED_ELAPSED_TIME'}, inplace=True)
df.rename(columns={'ELAPSED_TIME': 'ACTUAL_ELAPSED_TIME'}, inplace=True)
df.rename(columns={'FL_DATE': 'FLIGHT_DATE'}, inplace=True)
df.rename(columns={'FL_NUMBER': 'FLIGHT_NUMBER'}, inplace=True)

df.rename(columns={'WHEELS_OFF': 'WHEELS_LEFT_ORIGIN'}, inplace=True)
df.rename(columns={'WHEELS_ON': 'WHEELS_REACHED_DESTINATION'}, inplace=True)
df.rename(columns={'TAXI_IN': 'TAXI_TIME_DESTINATION'}, inplace=True)
df.rename(columns={'TAXI_OUT': 'TAXI_TIME_ORIGIN'}, inplace=True)
df.rename(columns={'AIR_TIME': 'IN_AIR_DURATION'}, inplace=True)

# Recheck
a simple recheck of all data statistics to see if the data is fine before donwloading

In [847]:
df.shape

(2999998, 37)

In [848]:
# Check for missing values
df.isnull().sum()

FLIGHT_DATE                   0
year                          0
quarter                       0
month                         0
day                           0
AIRLINE                       0
AIRLINE_CODE                  0
DOT_CODE                      0
FLIGHT_NUMBER                 0
ORIGIN                        0
ORIGIN_CITY                   0
DEST                          0
DEST_CITY                     0
SCHEDULED_DEP_TIME            0
ACTUAL_DEP_TIME               0
DEP_DELAY                     0
TAXI_TIME_ORIGIN              0
WHEELS_LEFT_ORIGIN            0
WHEELS_REACHED_DESTINATION    0
TAXI_TIME_DESTINATION         0
SCHEDULED_ARR_TIME            0
ACTUAL_ARR_TIME               0
ARR_DELAY                     0
CANCELLED                     0
CANCELLATION_CODE             0
DIVERTED                      0
SCHEDULED_ELAPSED_TIME        0
ACTUAL_ELAPSED_TIME           0
IN_AIR_DURATION               0
DISTANCE                      0
DELAY_DUE_CARRIER             0
DELAY_DU

In [849]:
# Display data types
df.dtypes

FLIGHT_DATE                   datetime64[ns]
year                                   int32
quarter                                int32
month                         string[python]
day                                    int32
AIRLINE                       string[python]
AIRLINE_CODE                  string[python]
DOT_CODE                               int64
FLIGHT_NUMBER                          int64
ORIGIN                        string[python]
ORIGIN_CITY                   string[python]
DEST                          string[python]
DEST_CITY                     string[python]
SCHEDULED_DEP_TIME                     int64
ACTUAL_DEP_TIME                      float64
DEP_DELAY                            float64
TAXI_TIME_ORIGIN                     float64
WHEELS_LEFT_ORIGIN                   float64
WHEELS_REACHED_DESTINATION           float64
TAXI_TIME_DESTINATION                float64
SCHEDULED_ARR_TIME                     int64
ACTUAL_ARR_TIME                      float64
ARR_DELAY 

In [850]:
df

FLIGHT_DATE  year  quarter month  day                 AIRLINE  \
0        2019-01-09  2019        1   Jan    9   United Air Lines Inc.   
1        2022-11-19  2022        4   Nov   19    Delta Air Lines Inc.   
2        2022-07-22  2022        3   Jul   22   United Air Lines Inc.   
3        2023-03-06  2023        1   Mar    6    Delta Air Lines Inc.   
4        2020-02-23  2020        1   Feb   23        Spirit Air Lines   
...             ...   ...      ...   ...  ...                     ...   
2999995  2022-11-13  2022        4   Nov   13  American Airlines Inc.   
2999996  2022-11-02  2022        4   Nov    2  American Airlines Inc.   
2999997  2022-09-11  2022        3   Sep   11    Delta Air Lines Inc.   
2999998  2019-11-13  2019        4   Nov   13        Republic Airline   
2999999  2019-06-15  2019        2   Jun   15  Southwest Airlines Co.   

        AIRLINE_CODE  DOT_CODE  FLIGHT_NUMBER ORIGIN          ORIGIN_CITY  \
0                 UA     19977           1562    FLL  Fort Lauderdale, FL   
1                 DL     19790           1149    MSP      Minneapolis, MN   
2                 UA     19977            459    DEN           Denver, CO   
3                 DL     19790           2295    MSP      Minneapolis, MN   
4                 NK     20416            407    MCO          Orlando, FL   
...              ...       ...            ...    ...                  ...   
2999995           AA     19805           1522    JAX     Jacksonville, FL   
2999996           AA     19805           1535    ORD          Chicago, IL   
2999997           DL     19790           2745    HSV       Huntsville, AL   
2999998           YX     20452           6134    BOS           Boston, MA   
2999999           WN     19393           2823    LGB       Long Beach, CA   

        DEST              DEST_CITY  SCHEDULED_DEP_TIME  ACTUAL_DEP_TIME  \
0        EWR             Newark, NJ                1155           1151.0   
1        SEA            Seattle, WA                2120           2114.0   
2        MSP        Minneapolis, MN                 954           1000.0   
3        SFO      San Francisco, CA                1609           1608.0   
4        DFW  Dallas/Fort Worth, TX                1840           1838.0   
...      ...                    ...                 ...              ...   
2999995  CLT          Charlotte, NC                1742           1740.0   
2999996  AUS             Austin, TX                1300           1254.0   
2999997  ATL            Atlanta, GA                 534            615.0   
2999998  LGA           New York, NY                1600           1555.0   
2999999  SJC           San Jose, CA                 730            727.0   

         DEP_DELAY  TAXI_TIME_ORIGIN  WHEELS_LEFT_ORIGIN  \
0             -4.0              19.0              1210.0   
1             -6.0               9.0              2123.0   
2              6.0              20.0              1020.0   
3             -1.0              27.0              1635.0   
4             -2.0              15.0              1853.0   
...            ...               ...                 ...   
2999995       -2.0              10.0              1750.0   
2999996       -6.0              10.0              1304.0   
2999997       41.0              16.0               631.0   
2999998       -5.0              19.0              1614.0   
2999999       -3.0               9.0               736.0   

         WHEELS_REACHED_DESTINATION  TAXI_TIME_DESTINATION  \
0                            1443.0                    4.0   
1                            2232.0                   38.0   
2                            1247.0                    5.0   
3                            1844.0                    9.0   
4                            2026.0                   14.0   
...                             ...                    ...   
2999995                      1845.0                    6.0   
2999996                      1514.0                    5.0   
2999997       

# Download cleaned dataset

In [851]:
# Save the cleaned DataFrame to a CSV file
df.to_csv('../data/flights_sample_3m_cleaned.csv', index=False)

In [852]:
# Extract first 1000 rows
df_1000 = df.head(1000)

# Save to CSV with column headers
df_1000.to_csv('../data/flights_sample_1k_cleaned.csv', index=False)

In [853]:
# revert pandas settings back to normal
pd.reset_option('display.max_columns')

# Last run

In [854]:
get_current_datetime()

'2025-05-28 15:17:25'